kratos/processes/time_averaging_process.cpp

```diff
@@ -0,0 +1,292 @@
+
+        std::stringstream msg;
+        msg << "Integrating historical";
+
+        for (const std::string& variable_name : mVariableNamesList)
+        {
+            if (KratosComponents<Variable<double>>::Has(variable_name))
+            {
+                const Variable<double>& r_variable =
+                    KratosComponents<Variable<double>>::Get(variable_name);
+                this->CalculateTimeIntegratedQuantity(r_nodes, r_variable, delta_time);
+            }
+            else if (KratosComponents<Variable<array_1d<double, 3>>>::Has(variable_name))
+            {
+                const Variable<array_1d<double, 3>>& r_variable =
+                    KratosComponents<Variable<array_1d<double, 3>>>::Get(variable_name);
+                this->CalculateTimeIntegratedQuantity(r_nodes, r_variable, delta_time);
+            }
+
+            msg << " " << variable_name << ",";
+        }
+
+        mCurrentTime += delta_time;
+
+        msg.seekp(-1, msg.cur);
+        msg << " variable(s) in " << mModelPartName << ".\n";
+
+        KRATOS_INFO_IF(this->Info(), mEchoLevel > 1) << msg.str();
+    }
+
+    KRATOS_CATCH("");
+}
+
+void TimeAveragingProcess::ExecuteFinalizeSolutionStep()
+{
+    Execute();
+}
+
+bool TimeAveragingProcess::IsIntegrationStep() const
+{
+    const ProcessInfo& r_process_info =
+        mrModel.GetModelPart(mModelPartName).GetProcessInfo();
+
+    if (KratosComponents<Variable<int>>::Has(mIntegrationControlVariableName))
+    {
+        const Variable<int>& r_variable =
+            KratosComponents<Variable<int>>::Get(mIntegrationControlVariableName);
+
+        KRATOS_ERROR_IF(!r_process_info.Has(r_variable))
+            << "\"" << mIntegrationControlVariableName
+            << "\" not found in process info of " << mModelPartName << ".\n";
+
+        return (r_process_info[r_variable] >=
+                mrParameters["integration_start_point_control_value"].GetInt());
+    }
+    else if (KratosComponents<Variable<double>>::Has(mIntegrationControlVariableName))
+    {
+        const Variable<double>& r_variable =
+            KratosComponents<Variable<double>>::Get(mIntegrationControlVariableName);
+
+        KRATOS_ERROR_IF(!r_process_info.Has(r_variable))
+            << "\"" << mIntegrationControlVariableName
+            << "\" not found in process info of " << mModelPartName << ".\n";
+
+        return (r_process_info[r_variable] >=
+                mrParameters["integration_start_point_control_value"].GetDouble());
+    }
+    else
+    {
+        return false;
+    }
+}
+
+std::string TimeAveragingProcess::Info() const
+{
+    return std::string("TimeAveragingProcess");
+}
+
+void TimeAveragingProcess::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << this->Info();
+}
+
+void TimeAveragingProcess::PrintData(std::ostream& rOStream) const
+{
+}
+
+template <typename TDataType>
+void TimeAveragingProcess::InitializeTimeAveragedQuantity(ModelPart::NodesContainerType& rNodes,
```

Comment:
I think this can be done using the variable_utils

Reply:
done :)